### PR TITLE
feat: typescript nextjs projects

### PR DESCRIPTION
**Pull request by Chriscbr**
*Issue #, if available:* N/A

*Description of changes:* Added a nextjs-ts project option which extends the typescript-app project class. Grouped some options together between the two nextjs project types, and implemented basic Next.js tsconfig project defaults so that both nextjs and nextjs-ts project types work out of the box.

----

Some notes:

I made a slight modification of TypeScriptConfig so that I can mark `tsconfig.json` as mutable, since the 'next build' command fails if it's immutable. The reason for this I believe is that the 'next' command tries to automatically change your tsconfig file if it sees that some options set incorrectly based on what Next.js's build system wants. I could send a feature request to the Next.js team to see if they can enable a CLI flag that skips this behavior, but for now this workaround makes the most sense.

There is a two-line file named `next-env.d.ts` that must be added to the root directory for TypeScript Next.js projects to successfully build. I included this in the PR, but it's worth suggesting that we should also eventually add an escape hatch so a user can omit or change the `next-env.d.ts` file if needed -- but I'm not sure what that mechanism should be, so I've left that unimplemented for now.

~Lastly, many Next.js projects don't have test directories from what I've seen, so I didn't include sample code for any. But this causes eslint to complain, and changing how the default eslint command behaves warrants a separate PR (see #182), and using `eslint: false` can lead to #180, so I ended up adding a small piece of cruft in the project constructor for now:~

```
this.addTestCommand('echo 0');
```

(I ended up overriding the default TypeScriptProject build command so this isn't an issue)

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Chriscbr** [comment]
Not totally sure why this build is failing (works fine locally) -- is it not possible to override the docstring of class?

File: src/typescript.ts [hunk 1]
```diff
@@ -503,6 +560,7 @@ export class TypescriptConfig {
   public readonly include: string[];
   public readonly exclude: string[];
   public readonly fileName: string;
+  public file: JsonFile;
```
Comment (eladb) [inline src/typescript.ts:563]:
```suggestion
  public readonly file: JsonFile;
```

File: src/typescript.ts [hunk 2]
```diff
@@ -380,6 +416,20 @@ export interface TypeScriptCompilerOptions {
    */
   readonly module?: string;
 
+  /**
+   * Determine how modules get resolved. Either "Node" for Node.js/io.js style resolution, or "Classic".
+   *
+   * @default 'node'
+   */
+  readonly moduleResolution?: string;
```
Comment (eladb) [inline src/typescript.ts:424]: enum?

File: src/typescript.ts [hunk 3]
```diff
@@ -365,6 +388,19 @@ export interface TypeScriptCompilerOptions {
...
+   * with transpileModule or @babel/plugin-transform-typescript) would be safe.
+   */
+  readonly isolatedModules?: boolean;
+
+  /**
+   * Support JSX in .tsx files: "react", "preserve", "react-native"
+   *
+   * @default undefined
+   */
+  readonly jsx?: string;
```
Comment (eladb) [inline src/typescript.ts:402]: enum?

File: src/typescript.ts [hunk 4]
```diff
@@ -365,6 +388,19 @@ export interface TypeScriptCompilerOptions {
    */
   readonly inlineSources?: boolean;
 
+  /**
+   * Perform additional checks to ensure that separate compilation (such as
+   * with transpileModule or @babel/plugin-transform-typescript) would be safe.
+   */
```
Comment (eladb) [inline src/typescript.ts:394]: @default

File: src/next.ts [hunk 5]
```diff
@@ -25,55 +21,239 @@ export interface NextJsProjectOptions extends NodeProjectOptions {
...
+        },
+        ...options.tsconfig,
+      },
+      ...options,
+
+      // never generate default TypeScript sample code, since this class provides its own
+      sampleCode: false,
+    });
+
+    this.srcDir = options.srcdir ?? 'pages';
```
Comment (eladb) [inline src/next.ts:154]: Kind of don't like the inconsistency `srcDir` versus `srcdir`. Let's normalize to `srcdir`?

File: src/next.ts [hunk 6]
```diff
@@ -25,55 +21,239 @@ export interface NextJsProjectOptions extends NodeProjectOptions {
...
-    this.addScripts({
-      dev: 'next dev',
-      build: 'next build',
-      server: 'next start',
+    // NextJS CLI commands, see: https://nextjs.org/docs/api-reference/cli
+    this.addScript('dev', 'next dev');
+    this.start?.addEntry('dev', {
+      desc: 'Starts the Next.js application in development mode',
+      category: StartEntryCategory.BUILD,
     });
```
Comment (eladb) [inline src/next.ts:169]: You could reuse this with the `NextProject` using `Component`s. You can create a `NextComponent` and then compose it here and there.

**Review by eladb — CHANGES_REQUESTED** [review]
minor-

**Review by eladb — APPROVED** [review]
🤗🤗🤗

**Pull request merged**